import/controls/permissionmodel.cpp

```diff
@@ -0,0 +1,92 @@
+/****************************************************************************
+**
+** Copyright (C) 2020 Open Mobile Platform LLC
+**
+****************************************************************************/
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "permissionmodel.h"
+#include "webengine.h"
+#include "permissionmanager.h"
+
+PermissionModel::PermissionModel(QObject *parent) : QAbstractListModel(parent)
+{
+    SailfishOS::WebEngine *webEngine = SailfishOS::WebEngine::instance();
+
+    connect(webEngine, &SailfishOS::WebEngine::recvObserve, this, &PermissionModel::handleRecvObserve);
+
+    webEngine->addObserver(QStringLiteral("embed:perms:all"));
+
+    PermissionManager::sendRequest("get-all");
+}
+
+QVariant PermissionModel::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid())
+        return QVariant();
+
+    switch (role) {
+    case Uri:
+        return m_permissionList[index.row()].m_uri;
+    case Type:
+        return m_permissionList[index.row()].m_type;
+    case Capability:
+        return m_permissionList[index.row()].m_capability;
+    default:
+        return QVariant();
+    }
+}
+
+int PermissionModel::rowCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent)
+    return m_permissionList.count();
+}
+
+QHash<int, QByteArray> PermissionModel::roleNames() const
+{
+    QHash<int, QByteArray> roles;
+    roles[Uri] = "uri";
+    roles[Type] = "type";
+    roles[Capability] = "capability";
+    return roles;
+}
+
+void PermissionModel::handleRecvObserve(const QString &message, const QVariant &data)
+{
+    if (message == "embed:perms:all") {
+        insertList(qvariant_cast<QVariantList>(data));
+    }
+}
+
+void PermissionModel::insertList(const QVariantList &data)
+{
+    beginResetModel();
+
+    m_permissionList.clear();
+
+    for (const auto &iter : data) {
+        QVariantMap varMap = iter.toMap();
+        m_permissionList.append(Permission(varMap.value("uri").toString(),
+                                           varMap.value("type").toString(),
+                                           varMap.value("capability").toString()));
+    }
+    endResetModel();
+}
+
+void PermissionModel::remove(int index)
+{
+    beginRemoveRows(QModelIndex(), index, index);
+    m_permissionList.removeAt(index);
```

Comment:
Maybe you should call `PermissionManager::remove()` otherwise the model just hides an entry.